pkg/server/podsubresourceproxy_handler.go

```diff
@@ -0,0 +1,235 @@
+			if !ok {
+				return nil, fmt.Errorf("expected synctarget to be of type %T, got %T", &workloadv1alpha1.SyncTarget{}, synctargets[0])
+			}
+			return synctarget, nil
+		},
+	}
+}
+
+type podSubresourceProxyHandler struct {
+	apiHandler                   http.Handler
+	podSubresourcePathRegex      *regexp.Regexp
+	getPodByName                 func(ctx context.Context, cluster logicalcluster.Name, namespace, podName string) (*corev1.Pod, error)
+	getSyncTargetBySynctargetKey func(ctx context.Context, synctargetKey string) (*workloadv1alpha1.SyncTarget, error)
+}
+
+func (b *podSubresourceProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	logger := klog.FromContext(req.Context())
+	cluster := request.ClusterFrom(req.Context())
+	if cluster.Name.Empty() {
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	// Validate that the call could be interesting to us...
+	if !strings.Contains(req.URL.Path, "/api/v1/namespaces/") {
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	// Apply regex
+	matches := b.podSubresourcePathRegex.FindStringSubmatch(req.URL.Path)
+	if len(matches) != 5 {
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	namespace := matches[1]
+	podName := matches[2]
+	subresource := matches[3]
+	arguments := matches[4]
+
+	// If everything is empty, just return..
+	if namespace == "" || podName == "" || subresource == "" {
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	// Now let's start the proxying
+	logger.Info("proxying pod subresource", "namespace", namespace, "podName", podName, "subresource", subresource, "arguments", arguments)
+
+	pod, err := b.getPodByName(req.Context(), cluster.Name, namespace, podName)
+	if apierrors.IsNotFound(err) {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewNotFound(schema.GroupResource{Group: "", Resource: "pods"}, podName),
+			errorCodecs, schema.GroupVersion{}, w, req,
+		)
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+	if err != nil {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewInternalError(err),
+			errorCodecs, schema.GroupVersion{}, w, req,
+		)
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	// Let's get the synctargetKey
+	var synctargetKey string
+	for k, v := range pod.GetLabels() {
+		if strings.HasPrefix(k, workloadv1alpha1.ClusterResourceStateLabelPrefix) {
+			if v == "Upsync" {
+				synctargetKey = strings.TrimPrefix(k, workloadv1alpha1.ClusterResourceStateLabelPrefix)
+				break
+			}
+		}
+	}
+	if synctargetKey == "" {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewBadRequest(fmt.Sprintf("pod %q is not upsynced", podName)),
+			errorCodecs, schema.GroupVersion{}, w, req,
+		)
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	synctarget, err := b.getSyncTargetBySynctargetKey(req.Context(), synctargetKey)
+	if apierrors.IsNotFound(err) {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewNotFound(schema.GroupResource{Group: "workload.kcp.dev", Resource: "synctargets"}, synctargetKey),
```

Comment:
Does this leak information the user would otherwise not be able to see?

Reply:
Fixed, PTAL!